### DISCUSSION TITLE
Dropping `allow_nil` in favor of using only `default`

### DISCUSSION THREAD

**Discussion by afuno** [sunny/actor#109] [Ideas]
Now there is confusion when using `allow_nil` and `default` at the same time. Or just using `allow_nil: true`.

Example:

```ruby
input :active,
      type: [TrueClass, FalseClass],
      default: false,
      allow_nil: false
```

```ruby
input :data, type: Hash, allow_nil: true
```

Both of these examples are incorrect and will result in an unexpected "is missing" error. It seems to me that these two checks conflict with each other.

I suggest the following solution:

- By default, all inputs and outputs are required. No need to write `allow_nil: false` anymore.
- Delete the `allow_nil` attribute from the public API.
- If it is necessary to allow nil, then use the `default` attribute.

@sunny What do you think about it? I would like to discuss this before starting the implementation of the prototype.

**Comment by sunny** [comment]
Hi @afuno!

> I suggest the following solution:
> - By default, all inputs and outputs are required. No need to write `allow_nil: false` anymore.

To clarify the discussion, I think we must first define what we call “required”.

To me *required* means that the argument must be provided, but can be `nil`. In the same way that in Ruby `def build(bear:); end` the `bear` argument is *required* as it raises when called without setting the `bear` argument but does not raise when called with `bear: nil`.

In that regard, all inputs are *already* required, except when a `default` is set. This is the same behavior as with Ruby keyword arguments.

> Both of these examples are incorrect and will result in an unexpected "is missing" error. It seems to me that these two checks conflict with each other.

Can you provide a test case for that? When trying this out, this is, to me, the expected behavior:

```rb
class BuildBear < Actor
  input :active,
        type: [TrueClass, FalseClass],
        default: false,
        allow_nil: false
end

BuildBear.call
# => no error since the value is `false`
BuildBear.call(active: true)
# => no error since the value is `true`
BuildBear.call(active: false)
# => no error since the value is `false`
BuildBear.call(active: nil)
# => ServiceActor::ArgumentError: The "active" input on "BuildBear" does not allow nil values
# The error here is expected because we *don’t* allow nil

class ScratchBark < Actor
  input :data, type: Hash, allow_nil: true
end

ScratchBark.call
# => ServiceActor::ArgumentError: The "data" input on "ScratchBark" is missing
# The error here is expected since the argument does not have a default.
ScratchBark.call(data: nil)
# => no error, since the argument is given and allows nil
```

> By default, all inputs and outputs are required.

If by “required”, you meant “should not accept nil”, I think that would be a major pain to upgrade apps that don’t rely on that behavior. I‘d much rather continue to allow `nil` as a default.

Perhaps we can remove the `allow_nil: false` option if it’s confusing, and perhaps ask people to use something like this instead:

```rb
input :data, type: Hash, must: { not_be_nil: -> value { !value.nil? } }
```

Or perhaps we can rename it to `must_not_be_nil: true` if it helps clarify.
